core/jvm/src/main/scala/scalaz/zio/IOApp.scala

```diff
@@ -44,7 +46,15 @@ trait IOApp extends RTS {
    * The Scala main function, intended to be called only by the Scala runtime.
    */
   final def main(args0: Array[String]): Unit =
-    unsafeRun(run(args0.toList)) match {
+    unsafeRun(
+      for {
+        fiber <- run(args0.toList).fork
+        _ = Runtime.getRuntime.addShutdownHook(new Thread {
```

Comment:
This should be changed to:

```scala
_ <- IO.sync(Runtime.......)
```

Otherwise it will run in-between binds in an effectful way that's not guaranteed to have any meaning.

Reply:
It seems like it should work correctly based on this example (Scala will create a Tuple with the fiber and the result of adding the hook):
```
scala> val (x, y, z) = (Some(1), Some(2), Some(3))
scala> desugar{
     |     for{
     |       i <- x
     |       _ = i
     |       k <- z
     |     } yield i + k
     |   }
$line3.$read.$iw.$iw.x.map[(Int, Int)](((i: Int) => {
  val x$1: Int = i;
  scala.Tuple2.apply[Int, Int](i, x$1)
})).flatMap[Int](((x$2: (Int, Int)) => (x$2: (Int, Int) @unchecked) match {
  case (_1: Int, _2: Int)(Int, Int)((i @ _), _) => $line3.$read.$iw.$iw.z.map[Int](((k: Int) => i.+(k)))
}))
```
However I do see that this code can be improved by your suggestion, so I will go ahead and create a new pull request